src/ipvs/ip_vs_proto_tcp.c

```diff
@@ -306,46 +306,129 @@ static void tcp_in_remove_ts(struct tcphdr *tcph)
-                for (i = 0; i < tcp_opt_len; i++) {
-                    *(ptr - 2 + i) = TCP_OPT_NOP;
+            if (opsize < 2)             /* silly options */
+                goto fini;
+            if (opsize > optlen)        /* partial options */
+                goto fini;
+            if (opcode == TCP_OPT_ADDR) {
+                fast += opsize;
+            } else {
+                for (i = 0; i < opsize; i++) {
+                    if (slow != fast)
+                        *slow = *fast;
+                    slow++;
+                    fast++;
                 }
-                /* DON'T RETURN
-                 * keep search other TCP_OPT_ADDR ,and clear them.
-                 * See https://github.com/iqiyi/dpvs/pull/925 for more detail. */
             }
 
             ptr += opsize - 2;
-            len -= opsize;
+            optlen -= opsize;
             break;
         }
     }
+
+fini:
+    pruned = payload - slow;
+    if (pruned < reqlen) {
+        /* further trim the options, the tcp functionality relies on unprotected
+         * options may get hurt, refer to:
+         *   https://www.iana.org/assignments/tcp-parameters/tcp-parameters.xhtml
+         *   #tcp-parameters-1
+         * */
+        ptr = slow;
+        slow = fast = (unsigned char *)(tcph + 1);
+        if (tcph->syn)
+            opts_protected = (1ULL << 2) | (1ULL << 3) | (1ULL << 4)    /* MSS, WS, SACKP */
+                | (1ULL << 8) | (1ULL << 30) | (1ULL << 34);            /* TS, MPTCP, TFO */
+        else
+            opts_protected = (1ULL << 8);    /* TS, drop SACK, MPTCP DSS/REMOVE_ADDR */
+        while (fast < ptr) {
+            opcode = *fast;
+            opsize = *(fast + 1);
+            if (opcode < 64 && ((1ULL << opcode) & opts_protected)
+                    && (opsize <= opts_maxlen[opcode])) {
+                for (i = 0; i < opsize; i++)
+                    *slow++ = *fast++;
+                opts_protected ^= (1ULL << opcode);
+            } else {
+                fast += opsize;
+                pruned += opsize;
+                if (pruned >= reqlen) {
+                    while (fast < ptr)
+                        *slow++ = *fast++;
+                    break;
+                }
+            }
+        }
+        pruned = payload - slow;
+    }
+    if (pruned > 0) {
+        while (pruned & 0x3) {  /* 4-bytes alignment for tcp options */
+            *slow++ = 0;
+            pruned--;
+        }
+        if (!pruned)
+            return 0;
+        /* trim the packet */
+        l3hdr = rte_pktmbuf_mtod(mbuf, void *);
+        if (unlikely(mbuf_may_pull(mbuf, mbuf->pkt_len) != 0)) {
+            memset(slow, 0, pruned);
+            return EDPVS_INVPKT;
+        }
+        if (unlikely(payload - l3hdr > mbuf->pkt_len)) {
+            memset(slow, 0, pruned);
+            return EDPVS_INVPKT;
+        }
+        memmove(slow, payload, mbuf->pkt_len - (payload - l3hdr));
```

Comment:
leave reqlen bytes for tcp_in_add_toa ?

Reply:
` tcp_in_prune_options` tries to find `reqlen`  bytes in tcp option space, and stops immediately after achieving the goal.  Then the memory after `slow` pointer in tcp option  is invalid and should be trimmed.

Reply:
Yes, reserving reqlen bytes spares one memory shift later in toa inserting. But I think it require changes in the `tcp_in_add_toa ` function. I'd not like to do more changes, so trimming all bytes released from tcp option here.